Vector Static layers CRUD

metemaddar opened the pull request:
ADD CRUD for Vector Static Layers

## Description

- [x] CREATE endpoing
- [x] READ endpoint
- [x] UPDATE endpoint
- [x] DELETE endpoint
- [x] fix READ endpoint
- [x] VALIDATIONS
    - [x] Simple upload data validation (while `geopandas.read_file()`)
- [x] TESTS

## Related Issue
 #1358

## Motivation and Context
From issue:

> In the schema extra, we are storing our static layers. These layers have no interactive logic but are used for visualization purposes only. For example, we show the number of accidents. We need an endpoint that allows us to upload this data via the dashboard.

## How Has This Been Tested? :test_tube: 
Passed pytests for:
- :heavy_check_mark: read one layer
- :heavy_check_mark: read list layers
- :heavy_check_mark: upload valid files
- :heavy_check_mark: upload invalid files
- :heavy_check_mark: delte static layer

## Screenshots (if appropriate):

## Actions
closes #1358
<!--- Put "closes #XXXX" in your comment to auto-close the issue that your PR fixes (if such). -->

<!--- Attribution: https://github.com/stevemao/github-issue-templates -->

EPajares commented:
I created a sample file that should fail. Maybe you can run what happens with the file when uploading it. 

[fail_file.zip](https://github.com/goat-community/goat/files/9066875/fail_file.zip)

metemaddar commented:
> I created a sample file that should fail. Maybe you can run what happens with the file when uploading it.
> 
> [fail_file.zip](https://github.com/goat-community/goat/files/9066875/fail_file.zip)

- The fail_file couldn't fail the test. Red by geopandas and uploaded to database.
- Also the sample file failed. Because the geo data was inside the sample directory at zip file. This should be fixed I think.

EPajares commented:
Ok sorry it seems QGIS added automatically the right projection when I created the file. Could you try again with this. Concerning the zipping. As for Shapefile we allows zipped file I think we should also allow it for the other formats. Not sure thouch if this causes any trouble in the backend?

This is the new invalid file.
[invali.zip](https://github.com/goat-community/goat/files/9069732/invali.zip)

metemaddar commented:
> This is the new invalid file.
> [invali.zip](https://github.com/goat-community/goat/files/9069732/invali.zip)

It added without any errors. Converted and added to database.

metemaddar commented on app/api/src/endpoints/v1/static_layers.py:
```diff
@@ -1,26 +1,101 @@
+
+
+@router2.post("/static")
+async def uplaod_static_layer(
```
### Auto Deleting UploadFile
Seems the UploadFile will be deleted after closing the session automatically as it is a SpooledTemporaryFile [ref](https://fastapi.tiangolo.com/tutorial/request-files/#uploadfile). I'm not sure about when the file exceeds it's max_size. Because then it will be written to disk. [Someone says](https://stackoverflow.com/a/59393654/1951027) the file will be deleted even it was written to disk. I didn't see any reference inside the [docs](https://docs.python.org/3/library/tempfile.html#tempfile.SpooledTemporaryFile). Maybe we should test or check more docs.

EPajares replied:
This is great!

metemaddar replied:
### Opening different file types inside zip file
The GeoPandas now can open different Geo data files inside zip file.
But the problem about zip files is that if the files are under a directory inside zip file the GeoPandas doesn't open them automatically, even if there is only one directory. It can determine the directory if we specify the directory name for it.
As for example in the sample.zip, It couldn't open it. In such situations, if we automatically detect the directory name and specify it for read_file() then it would open it.
I'm going to check the [documentation](https://geopandas.org/en/stable/docs/user_guide/io.html) for better solution.

EPajares replied:
Ok, great thanks! An alternative would be reading the Zipfile with another library and then unpacking it. Then it could be read with Geopandas, I guess.

metemaddar replied:
The only way I think is to:
1. detect if the zip file contains only one directory
2. save it to disk
3. open it by geopandas.read_file() by giving the sub directory.
4. remove the file.

metemaddar replied:
It should be like this I think:
![Open Uploaded PostGIS File](https://user-images.githubusercontent.com/3726265/181187410-84ab0612-addd-4a7c-9f01-112d33c82ca6.png)

EPajares replied:
Thanks for this great flow chart. I believe this is exactly how we should do it.

metemaddar replied:
:heavy_check_mark: Done
If the file type is not zip, it won't try to save file to disk. It uses the default functionality of SpooledTemporaryFile as it will be destroyed automatically.

EPajares replied:
Makes sense!

metemaddar commented on app/api/src/endpoints/v1/static_layers.py:
```diff
@@ -1,26 +1,101 @@
+    current_user: models.User = Depends(deps.get_current_active_superuser),
+    upload_file: UploadFile,
+):
+    try:
```
:heavy_check_mark:  Could encapsulate the validations for `data_frame` and this function could be updated for other validations. But for uploading data to database we should pass several data which can be same as another encapsulation function, so left unchanged.

metemaddar commented:
Files and Validations:
:white_check_mark:  single_directory.zip: /sample/shapefiles ...
:white_check_mark: no_directory.zip (zip with files directly compressed)
:white_check_mark: geojson.zip
:white_check_mark:  dpdData.**geojson**
:x: multiple_directories.zip (zip with multiple directories inside it)
:x: invalid.zip (As CSR is None or is 'unknown')
